### DISCUSSION TITLE
@variants directive overrides default styles

### DISCUSSION THREAD

**Discussion by maltebaer** [tailwindlabs/tailwindcss#5764] [Help] [answered]
Hey folks 👋🏽,

I'm trying to add a component to my codebase by adding a class to the components layer. Since I would like to have some state modifiers, let's say `hover`, I use the new `@variants` directive following the [documentation](https://tailwindcss.com/docs/extracting-components#extracting-component-classes-with-apply).

However, I cannot get it to work since the styles applied via the directive overrides my default style.

Since this is such a basic feature, I guess it must be some kind of error on my side. (BTW: using the old syntax works just fine.)

Check out the example: https://play.tailwindcss.com/hWQI7PqWTW?file=css
```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .btn {
    @apply bg-red-400 text-white px-3 py-5 rounded;
  }

  @variants hover {
    .btn {
      @apply bg-blue-400;
    }
  }

  /* .btn:hover {
    @apply bg-blue-400;
  } */
}

```
```html
<div class="p-8">
  <button class="btn">Click me</button>
  <button class="bg-red-400 text-white px-3 py-5 rounded hover:bg-blue-400">Click me</button>
</div>
```

**Comment by NovusOpus** [comment, marked as answer]
**Edit with the right Answer so people don't have to go all the way down the reply's for the answer:** 

```CSS
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .btn {
    @apply bg-red-400 text-white px-3 py-5 rounded hover:bg-blue-400 focus:ring active:bg-red-500;
  }
}
```


**Original:** 
The code you have commented seems a better approach to me, less code more clean:

```CSS
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .btn {
    @apply bg-red-400 text-white px-3 py-5 rounded;
  }
  .btn:hover {
    @apply bg-blue-400;
  }
}
```
But since you want to know what's wrong with your current approach following the documentation, In that case I'll leave that to a **Tailwind Staff**, so they can reply  to you properly.  

I'm guessing why it's not working, but I could be wrong. Sorry if my reply may be a bit pointless.

**Reply by maltebaer** [reply]
Hey @novuspace thanks for your answer.

Yep, the commented part is what I'm used to (and in that scenario totally fine).

However, this is just a simplified version. Actually I want to add the modification to multiple states. In that case the new syntax comes in handy, since you don't need to repeat yourself.

See
```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .btn {
    @apply bg-red-400 text-white px-3 py-5 rounded;
  }

  .btn:hover {
    @apply bg-blue-400;
  }
  .btn:focus {
    @apply bg-blue-400;
  }
  .btn:active {
    @apply bg-blue-400;
  }
}
```
opposed to
```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .btn {
    @apply bg-red-400 text-white px-3 py-5 rounded;
  }

 @variants hover, focus, active {
  .btn {
    @apply bg-blue-400;
  }
}
 ```

**Reply by NovusOpus** [reply]
@maltebaer Oh I see, in that case my guess is: You also have to include the `hover:, focus:, active:` in the `@apply`, like this:

```CSS
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .btn {
    @apply bg-red-400 text-white px-3 py-5 rounded;
  }

  @variants hover, focus, active {
  .btn {
    @apply hover:bg-blue-400 focus:ring focus:ring-red-700 active:bg-red-500;
  }
}
}
```
So please **Tailwind Staff** (@simonswiss or @adamwathan). Correct me if I'm wrong, Or if there's a better way to do it .

Update: So I tried it with **Tailwind Play** seems to work fine, even better :

```CSS
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .btn {
    @apply bg-red-400 text-white px-3 py-5 rounded hover:bg-blue-400 focus:ring  active:bg-red-500;
  }
}
```
This also seems to work exactly like the first code.

**Reply by maltebaer** [reply]
Hey @novuspace I just realised that I've mixed up two concepts. The `@variants` directive is for creating variant classes of your own custom utilities/components etc and not for applying different styles at different states.

So I bet your suggestion is the right way to do it.
```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .btn {
    @apply bg-red-400 text-white px-3 py-5 rounded hover:bg-blue-400 focus:ring active:bg-red-500;
  }
}
```

Thanks a lot to pointing me into this direction!

**Reply by NovusOpus** [reply]
@maltebaer Sure, glad to help, Also was fun to play with it, And I learned something new, Enjoying Tailwind Even more every day.
